File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/play/shops.ts
/**
 * DM-managed settlement shops with deterministic stock/prices and player
 * buy/sell operations against campaign inventory and currency. See
 * shared.ts for the ownership model and settlements.ts for settlements.
 */

import { getDb } from '../../db.ts';
import type { ApiResult, JsonValue } from '../../types.ts';
import { isValidIntInRange } from '../../validation.ts';
import { VALID_ITEM_IDS } from './inventory.ts';
import {
  authenticate,
  isActor,
  isApiResult,
  findCampaign,
  requireParticipant,
  resolveCharacterOwner,
  findMemberByCharacterId,
} from './shared.ts';
import { findSettlement } from './settlements.ts';

type ShopRow = {
  shop_id: string;
  name: string;
  stock_json: string;
  buy_price: number;
  sell_price: number;
};

function shopBody(row: ShopRow): JsonValue {
  return {
    shop_id: row.shop_id,
    name: row.name,
    stock: JSON.parse(row.stock_json),
    buy_price: row.buy_price,
    sell_price: row.sell_price,
  } as JsonValue;
}

function parseStock(value: unknown): Record<string, number> | ApiResult {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    return { status: 400, body: { error: 'stock must be a non-empty object' } };
  }
  const entries = Object.entries(value as Record<string, unknown>);
  if (entries.length === 0) {
    return { status: 400, body: { error: 'stock must be a non-empty object' } };
  }
  const stock: Record<string, number> = {};
  for (const [itemId, quantity] of entries) {
    if (!VALID_ITEM_IDS.has(itemId)) {
      return { status: 400, body: { error: 'stock keys must be valid catalog item ids' } };
    }
    if (!isValidIntInRange(quantity, 1, Number.MAX_SAFE_INTEGER)) {
      return { status: 400, body: { error: 'stock values must be positive integers' } };
    }
    stock[itemId] = quantity as number;
  }
  return stock;
}

function parseShopFields(
  body: JsonValue,
): { name: string; stock: Record<string, number>; buyPrice: number; sellPrice: number } | ApiResult {
  const name = body.name;
  if (typeof name !== 'string' || name.length === 0) {
    return { status: 400, body: { error: 'name must be a non-empty string' } };
  }

  const stock = parseStock(body.stock);
  if (isApiResult(stock)) return stock;

  if (!isValidIntInRange(body.buy_price, 1, Number.MAX_SAFE_INTEGER)) {
    return { status: 400, body: { error: 'buy_price must be a positive integer' } };
  }
  const buyPrice = body.buy_price as number;

  if (!isValidIntInRange(body.sell_price, 0, Number.MAX_SAFE_INTEGER)) {
    return { status: 400, body: { error: 'sell_price must be a nonnegative integer' } };
  }
  const sellPrice = body.sell_price as number;

  return { name, stock, buyPrice, sellPrice };
}

function nextShopSequence(db: ReturnType<typeof getDb>, campaignId: string, settlementId: string): number {
  const row = db
    .prepare(
      'SELECT COALESCE(MAX(sequence), 0) AS max_sequence FROM play_campaign_shops WHERE campaign_id = ? AND settlement_id = ?',
    )
    .get(campaignId, settlementId) as { max_sequence: number };
  return row.max_sequence + 1;
}

function findShop(
  db: ReturnType<typeof getDb>,
  campaignId: string,
  settlementId: string,
  shopId: string,
): ShopRow | ApiResult {
  const row = db
    .prepare(
      'SELECT shop_id, name, stock_json, buy_price, sell_price FROM play_campaign_shops WHERE campaign_id = ? AND settlement_id = ? AND shop_id = ?',
    )
    .get(campaignId, settlementId, shopId) as ShopRow | undefined;
  return row ?? { status: 404, body: { error: 'shop not found' } };
}

function hasDiscovered(
  db: ReturnType<typeof getDb>,
  campaignId: string,
  settlementId: string,
  characterId: string,
): boolean {
  return (
    db
      .prepare(
        'SELECT character_id FROM play_campaign_settlement_discoveries WHERE campaign_id = ? AND settlement_id = ? AND character_id = ?',
      )
      .get(campaignId, settlementId, characterId) !== undefined
  );
}

export function createShop(
  authHeader: string | undefined,
  campaignId: string,
  settlementId: string,
  body: JsonValue,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  if (campaign.owner !== actor.username) {
    return { status: 403, body: { error: 'only the campaign dm may create shops' } };
  }

  const settlement = findSettlement(db, campaignId, settlementId);
  if (isApiResult(settlement)) return settlement;

  const shopId = body.shop_id;
  if (typeof shopId !== 'string' || shopId.length === 0) {
    return { status: 400, body: { error: 'shop_id must be a non-empty string' } };
  }

  const parsed = parseShopFields(body);
  if (isApiResult(parsed)) return parsed;

  const existing = db
    .prepare('SELECT shop_id FROM play_campaign_shops WHERE campaign_id = ? AND settlement_id = ? AND shop_id = ?')
    .get(campaignId, settlementId, shopId);
  if (existing) {
    return { status: 409, body: { error: 'shop_id already exists in this settlement' } };
  }

  const sequence = nextShopSequence(db, campaignId, settlementId);
  const stockJson = JSON.stringify(parsed.stock);
  db.prepare(
    'INSERT INTO play_campaign_shops (campaign_id, settlement_id, shop_id, sequence, name, stock_json, buy_price, sell_price) VALUES (?, ?, ?, ?, ?, ?, ?, ?)',
  ).run(campaignId, settlementId, shopId, sequence, parsed.name, stockJson, parsed.buyPrice, parsed.sellPrice);

  return {
    status: 201,
    body: shopBody({
      shop_id: shopId,
      name: parsed.name,
      stock_json: stockJson,
      buy_price: parsed.buyPrice,
      sell_price: parsed.sellPrice,
    }),
  };
}

export function getShop(
  authHeader: string | undefined,
  campaignId: string,
  settlementId: string,
  shopId: string,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const settlement = findSettlement(db, campaignId, settlementId);
  if (isApiResult(settlement)) return settlement;

  const shop = findShop(db, campaignId, settlementId, shopId);
  if (isApiResult(shop)) return shop;

  if (campaign.owner === actor.username) {
    return { status: 200, body: shopBody(shop) };
  }

  const memberRow = db
    .prepare('SELECT character_id FROM play_campaign_members WHERE campaign_id = ? AND username = ?')
    .get(campaignId, actor.username) as { character_id: string } | undefined;
  if (!memberRow || !hasDiscovered(db, campaignId, settlementId, memberRow.character_id)) {
    return { status: 404, body: { error: 'shop not found' } };
  }

  return { status: 200, body: shopBody(shop) };
}

function parseTrade(body: JsonValue): { characterId: string; itemId: string; quantity: number } | ApiResult {
  const characterId = body.character_id;
  if (typeof characterId !== 'string' || characterId.length === 0) {
    return { status: 400, body: { error: 'character_id must be a non-empty string' } };
  }

  const itemId = body.item_id;
  if (typeof itemId !== 'string' || !VALID_ITEM_IDS.has(itemId)) {
    return { status: 400, body: { error: 'item_id must be a valid catalog item' } };
  }

  if (!isValidIntInRange(body.quantity, 1, Number.MAX_SAFE_INTEGER)) {
    return { status: 400, body: { error: 'quantity must be a positive integer' } };
  }

  return { characterId, itemId, quantity: body.quantity as number };
}

export function buyFromShop(
  authHeader: string | undefined,
  campaignId: string,
  settlementId: string,
  shopId: string,
  body: JsonValue,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const settlement = findSettlement(db, campaignId, settlementId);
  if (isApiResult(settlement)) return settlement;

  const shop = findShop(db, campaignId, settlementId, shopId);
  if (isApiResult(shop)) return shop;

  const trade = parseTrade(body);
  if (isApiResult(trade)) return trade;
  const { characterId, itemId, quantity } = trade;

  const owner = resolveCharacterOwner(db, campaignId, characterId);
  if (!owner) {
    return { status: 404, body: { error: 'character not found' } };
  }
  if (owner !== actor.username) {
    return { status: 403, body: { error: 'only the character owner may buy from a shop' } };
  }

  const member = findMemberByCharacterId(db, campaignId, characterId);
  if (isApiResult(member)) return member;

  const stock = JSON.parse(shop.stock_json) as Record<string, number>;
  const heldStock = stock[itemId] ?? 0;
  const cost = shop.buy_price * quantity;

  if (heldStock < quantity || member.gold < cost) {
    return { status: 409, body: { error: 'insufficient stock or funds' } };
  }

  const remainingStock = heldStock - quantity;
  stock[itemId] = remainingStock;
  db.prepare('UPDATE play_campaign_shops SET stock_json = ? WHERE campaign_id = ? AND settlement_id = ? AND shop_id = ?').run(
    JSON.stringify(stock),
    campaignId,
    settlementId,
    shopId,
  );

  const remainingGold = member.gold - cost;
  db.prepare('UPDATE play_campaign_members SET gold = ? WHERE campaign_id = ? AND character_id = ?').run(
    remainingGold,
    campaignId,
    characterId,
  );

  const existingStack = db
    .prepare('SELECT quantity FROM play_campaign_inventory_stacks WHERE campaign_id = ? AND character_id = ? AND item_id = ?')
    .get(campaignId, characterId, itemId) as { quantity: number } | undefined;
  const totalQuantity = (existingStack?.quantity ?? 0) + quantity;
  if (existingStack) {
    db.prepare(
      'UPDATE play_campaign_inventory_stacks SET quantity = ? WHERE campaign_id = ? AND character_id = ? AND item_id = ?',
    ).run(totalQuantity, campaignId, characterId, itemId);
  } else {
    db.prepare(
      'INSERT INTO play_campaign_inventory_stacks (campaign_id, character_id, item_id, quantity) VALUES (?, ?, ?, ?)',
    ).run(campaignId, characterId, itemId, totalQuantity);
  }

  return {
    status: 200,
    body: { character_id: characterId, item_id: itemId, quantity, gold: remainingGold, stock: remainingStock },
  };
}

export function sellToShop(
  authHeader: string | undefined,
  campaignId: string,
  settlementId: string,
  shopId: string,
  body: JsonValue,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const settlement = findSettlement(db, campaignId, settlementId);
  if (isApiResult(settlement)) return settlement;

  const shop = findShop(db, campaignId, settlementId, shopId);
  if (isApiResult(shop)) return shop;

  const trade = parseTrade(body);
  if (isApiResult(trade)) return trade;
  const { characterId, itemId, quantity } = trade;

  const owner = resolveCharacterOwner(db, campaignId, characterId);
  if (!owner) {
    return { status: 404, body: { error: 'character not found' } };
  }
  if (owner !== actor.username) {
    return { status: 403, body: { error: 'only the character owner may sell to a shop' } };
  }

  const member = findMemberByCharacterId(db, campaignId, characterId);
  if (isApiResult(member)) return member;

  const heldItem = db
    .prepare('SELECT quantity FROM play_campaign_inventory_stacks WHERE campaign_id = ? AND character_id = ? AND item_id = ?')
    .get(campaignId, characterId, itemId) as { quantity: number } | undefined;
  const heldQuantity = heldItem?.quantity ?? 0;

  if (heldQuantity < quantity) {
    return { status: 409, body: { error: 'insufficient inventory' } };
  }

  const remainingQuantity = heldQuantity - quantity;
  if (remainingQuantity === 0) {
    db.prepare('DELETE FROM play_campaign_inventory_stacks WHERE campaign_id = ? AND character_id = ? AND item_id = ?').run(
      campaignId,
      characterId,
      itemId,
    );
  } else {
    db.prepare(
      'UPDATE play_campaign_inventory_stacks SET quantity = ? WHERE campaign_id = ? AND character_id = ? AND item_id = ?',
    ).run(remainingQuantity, campaignId, characterId, itemId);
  }

  const proceeds = shop.sell_price * quantity;
  const totalGold = member.gold + proceeds;
  db.prepare('UPDATE play_campaign_members SET gold = ? WHERE campaign_id = ? AND character_id = ?').run(
    totalGold,
    campaignId,
    characterId,
  );

  const stock = JSON.parse(shop.stock_json) as Record<string, number>;
  const updatedStock = (stock[itemId] ?? 0) + quantity;
  stock[itemId] = updatedStock;
  db.prepare('UPDATE play_campaign_shops SET stock_json = ? WHERE campaign_id = ? AND settlement_id = ? AND shop_id = ?').run(
    JSON.stringify(stock),
    campaignId,
    settlementId,
    shopId,
  );

  return {
    status: 200,
    body: { character_id: characterId, item_id: itemId, quantity, gold: totalGold, stock: updatedStock },
  };
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T072004Z_claude_claude-sonnet-5_ruby-sinatra/routes/play/shops.rb
# DM-managed settlement shops with deterministic stock/prices, and player
# buy/sell operations backed by campaign inventory and currency.

def play_shop_payload(shop)
  {
    shop_id: shop['shop_id'],
    name: shop['name'],
    stock: JSON.parse(shop['stock_json']),
    buy_price: shop['buy_price'],
    sell_price: shop['sell_price']
  }
end

# Validates the create payload and returns the normalized
# [shop_id, name, stock, buy_price, sell_price] tuple.
def validate_shop_fields!(body)
  shop_id = body['shop_id']
  name = body['name']
  stock = body['stock']
  buy_price = body['buy_price']
  sell_price = body['sell_price']

  halt 400, { error: 'invalid shop_id' }.to_json unless shop_id.is_a?(String) && !shop_id.empty?
  halt 400, { error: 'invalid name' }.to_json unless name.is_a?(String) && !name.empty?
  halt 400, { error: 'invalid stock' }.to_json unless stock.is_a?(Hash) && !stock.empty?

  normalized_stock = {}
  stock.each do |item_id, quantity|
    halt 400, { error: 'invalid stock' }.to_json unless VALID_INVENTORY_ITEM_IDS.include?(item_id)
    halt 400, { error: 'invalid stock' }.to_json unless integerish(quantity) && quantity.to_i.positive?

    normalized_stock[item_id] = quantity.to_i
  end

  halt 400, { error: 'invalid buy_price' }.to_json unless integerish(buy_price) && buy_price.to_i.positive?
  halt 400, { error: 'invalid sell_price' }.to_json unless integerish(sell_price) && sell_price.to_i >= 0

  [shop_id, name, normalized_stock, buy_price.to_i, sell_price.to_i]
end

def find_play_shop!(campaign_id, settlement_id, shop_id)
  shop = db.execute(
    'SELECT * FROM play_shops WHERE campaign_id = ? AND settlement_id = ? AND shop_id = ?',
    [campaign_id, settlement_id, shop_id]
  ).first
  halt 404, { error: 'shop not found' }.to_json unless shop
  shop
end

post '/v1/play/campaigns/:id/settlements/:settlement_id/shops' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_owner!(campaign, user)

  settlement = find_play_settlement!(campaign['id'], params[:settlement_id])

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  shop_id, name, stock, buy_price, sell_price = validate_shop_fields!(body)

  existing = db.execute(
    'SELECT 1 FROM play_shops WHERE campaign_id = ? AND settlement_id = ? AND shop_id = ?',
    [campaign['id'], settlement['settlement_id'], shop_id]
  ).first
  halt 409, { error: 'shop already exists' }.to_json if existing

  db.execute(
    'INSERT INTO play_shops (campaign_id, settlement_id, shop_id, name, stock_json, buy_price, sell_price) VALUES (?, ?, ?, ?, ?, ?, ?)',
    [campaign['id'], settlement['settlement_id'], shop_id, name, stock.to_json, buy_price, sell_price]
  )

  status 201
  play_shop_payload(find_play_shop!(campaign['id'], settlement['settlement_id'], shop_id)).to_json
end

get '/v1/play/campaigns/:id/settlements/:settlement_id/shops/:shop_id' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  is_owner = require_play_participant!(campaign, user, 'not a campaign member')

  settlement = find_play_settlement!(campaign['id'], params[:settlement_id])
  shop = find_play_shop!(campaign['id'], settlement['settlement_id'], params[:shop_id])

  unless is_owner
    member = db.execute(
      'SELECT character_id FROM play_campaign_members WHERE campaign_id = ? AND username = ?',
      [campaign['id'], user['username']]
    ).first
    character_id = member && member['character_id']
    discovered_by = JSON.parse(settlement['discovered_by_json'])

    halt 404, { error: 'shop not found' }.to_json unless character_id && discovered_by.include?(character_id)
  end

  play_shop_payload(shop).to_json
end

post '/v1/play/campaigns/:id/settlements/:settlement_id/shops/:shop_id/buy' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  settlement = find_play_settlement!(campaign['id'], params[:settlement_id])
  shop = find_play_shop!(campaign['id'], settlement['settlement_id'], params[:shop_id])

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  character_id = body['character_id']
  item_id = body['item_id']
  quantity = body['quantity']

  halt 400, { error: 'invalid character_id' }.to_json unless character_id.is_a?(String) && !character_id.empty?

  member = find_play_member_by_character!(campaign['id'], character_id)
  owner = play_character_owner(campaign['id'], character_id, member)
  halt 403, { error: 'not the character owner' }.to_json unless owner == user['username']

  halt 400, { error: 'invalid item_id' }.to_json unless VALID_INVENTORY_ITEM_IDS.include?(item_id)
  halt 400, { error: 'invalid quantity' }.to_json unless integerish(quantity) && quantity.to_i.positive?
  quantity = quantity.to_i

  stock = JSON.parse(shop['stock_json'])
  available = stock[item_id] || 0
  halt 409, { error: 'insufficient stock' }.to_json if quantity > available

  cost = shop['buy_price'] * quantity

  gold_row = db.execute(
    'SELECT gold FROM play_character_gold WHERE campaign_id = ? AND character_id = ?',
    [campaign['id'], character_id]
  ).first
  gold = gold_row ? gold_row['gold'] : 0
  halt 409, { error: 'insufficient gold' }.to_json if cost > gold

  gold_after = gold - cost
  stock_after = available - quantity
  stock[item_id] = stock_after

  existing_item = db.execute(
    'SELECT quantity FROM play_character_inventory_items WHERE campaign_id = ? AND character_id = ? AND item_id = ?',
    [campaign['id'], character_id, item_id]
  ).first
  item_total = (existing_item ? existing_item['quantity'] : 0) + quantity

  db.execute(
    'INSERT INTO play_character_gold (campaign_id, character_id, gold) VALUES (?, ?, ?) ' \
    'ON CONFLICT (campaign_id, character_id) DO UPDATE SET gold = excluded.gold',
    [campaign['id'], character_id, gold_after]
  )
  db.execute(
    'INSERT INTO play_character_inventory_items (campaign_id, character_id, item_id, quantity) VALUES (?, ?, ?, ?) ' \
    'ON CONFLICT (campaign_id, character_id, item_id) DO UPDATE SET quantity = excluded.quantity',
    [campaign['id'], character_id, item_id, item_total]
  )
  db.execute(
    'UPDATE play_shops SET stock_json = ? WHERE campaign_id = ? AND settlement_id = ? AND shop_id = ?',
    [stock.to_json, campaign['id'], settlement['settlement_id'], shop['shop_id']]
  )

  {
    character_id: character_id,
    item_id: item_id,
    quantity: quantity,
    gold: gold_after,
    stock: stock_after
  }.to_json
end

post '/v1/play/campaigns/:id/settlements/:settlement_id/shops/:shop_id/sell' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  settlement = find_play_settlement!(campaign['id'], params[:settlement_id])
  shop = find_play_shop!(campaign['id'], settlement['settlement_id'], params[:shop_id])

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  character_id = body['character_id']
  item_id = body['item_id']
  quantity = body['quantity']

  halt 400, { error: 'invalid character_id' }.to_json unless character_id.is_a?(String) && !character_id.empty?

  member = find_play_member_by_character!(campaign['id'], character_id)
  owner = play_character_owner(campaign['id'], character_id, member)
  halt 403, { error: 'not the character owner' }.to_json unless owner == user['username']

  halt 400, { error: 'invalid item_id' }.to_json unless VALID_INVENTORY_ITEM_IDS.include?(item_id)
  halt 400, { error: 'invalid quantity' }.to_json unless integerish(quantity) && quantity.to_i.positive?
  quantity = quantity.to_i

  existing_item = db.execute(
    'SELECT quantity FROM play_character_inventory_items WHERE campaign_id = ? AND character_id = ? AND item_id = ?',
    [campaign['id'], character_id, item_id]
  ).first
  held = existing_item ? existing_item['quantity'] : 0
  halt 409, { error: 'insufficient inventory' }.to_json if quantity > held

  proceeds = shop['sell_price'] * quantity

  gold_row = db.execute(
    'SELECT gold FROM play_character_gold WHERE campaign_id = ? AND character_id = ?',
    [campaign['id'], character_id]
  ).first
  gold = gold_row ? gold_row['gold'] : 0
  gold_after = gold + proceeds

  held_after = held - quantity

  stock = JSON.parse(shop['stock_json'])
  stock_after = (stock[item_id] || 0) + quantity
  stock[item_id] = stock_after

  if held_after.zero?
    db.execute(
      'DELETE FROM play_character_inventory_items WHERE campaign_id = ? AND character_id = ? AND item_id = ?',
      [campaign['id'], character_id, item_id]
    )
  else
    db.execute(
      'UPDATE play_character_inventory_items SET quantity = ? WHERE campaign_id = ? AND character_id = ? AND item_id = ?',
      [held_after, campaign['id'], character_id, item_id]
    )
  end

  db.execute(
    'INSERT INTO play_character_gold (campaign_id, character_id, gold) VALUES (?, ?, ?) ' \
    'ON CONFLICT (campaign_id, character_id) DO UPDATE SET gold = excluded.gold',
    [campaign['id'], character_id, gold_after]
  )
  db.execute(
    'UPDATE play_shops SET stock_json = ? WHERE campaign_id = ? AND settlement_id = ? AND shop_id = ?',
    [stock.to_json, campaign['id'], settlement['settlement_id'], shop['shop_id']]
  )

  {
    character_id: character_id,
    item_id: item_id,
    quantity: quantity,
    gold: gold_after,
    stock: stock_after
  }.to_json
end
